### PR TITLE
Multi Adapter Value Function 

### PR DESCRIPTION
**Pull request by mnoukhov**
PR that extends #373 to also make the value function a separate model. Specifically, add a new set of LoRA parameters for the value function and learn them completely separately from the policy instead of making the value function a separate head on the policy. By default, also initialize the value function with the policy parameters to bootstrap it.

The upside is the value function is not affected by the policy and alignment tax should be reduced. The downside is you need two full forward passes to get the policy and value logits. 

Current issues
- I implemented this in StackLLaMA but #373 uses a different example
- Can't be merged into #373 because I rebased on the main branch
- Getting an error `path/to/torch/utils/checkpoint.py:31: UserWarning: None of the inputs have requires_grad=True. Gradients will be None` that is also present in #373, I believe

### CODE REVIEW

**Comment by mnoukhov** [comment]
Yep that's what I was thinking too. Once your branch is moved, I'll edit this to change the base branch to main and change from draft to real PR. I just started this as a draft and set your branch as the base branch to make my changes more clear i.e. setting `requires_grad=False` in https://github.com/lvwerra/trl/pull/428/files#diff-be4f28845d341065eb254da5db9fa8fee2e6b5347071ef98448f27acd70fe8a7R461-R463. 

But looking at the Files Changed actually doesn't correctly show the diff between my branch and yours for some reason. Let me know if you'd prefer if I rebased on your branch or did something else to make this draft more useful
